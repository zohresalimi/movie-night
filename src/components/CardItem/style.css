.card .tools-wrapper {
  font-size: 14px;
  position: absolute;
  z-index: 10;
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  top: 15px;
  left: 15px;
  background-color: #fff;
  border-radius: 2px;
}

.card .tools-wrapper div {
  padding: 10px 5px;
}

.card .tools-wrapper .ui.basic.buttons {
  box-shadow: none;
  border: 0;
  padding: 0;
}

/* .card .tools-wrapper .ui.icon.button.isFavorite {
  color: #fd6060;
} */

.card-wrapper .ui.message {
  position: fixed;
  top: 10px;
  left: 25%;
  z-index: 10;
}
