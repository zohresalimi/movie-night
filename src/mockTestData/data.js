import React, { useReducer } from "react";
import AppContext from "../store/context";
import reducer from "../store/reducer";

export const getTestState = () => {
  return {
    activePage: 1,
    movies: {
      1: {
        page: 1,
        total_pages: 2,
        total_results: 2,
        results: [
          {
            id: 10523,
            adult: false,
            poster_path: "/nJm8UF4zbUYEoku9seEtz1PTJGS.jpg",
          },
          {
            id: 50647,
            adult: false,
            poster_path: "/PlzkOA19mxEit7tV3AUnBYZ0io.jpg",
          },
        ],
      },
    },
    watchLaterList: {
      10523: {
        id: 10523,
        adult: false,
        poster_path: "/nJm8UF4zbUYEoku9seEtz1PTJGS.jpg",
      },
    },
    favoriteList: {
      10523: {
        id: 10523,
        adult: false,
        poster_path: "/nJm8UF4zbUYEoku9seEtz1PTJGS.jpg",
      },
    },
    apiConfig: {
      secure_base_url: "https://image.tmdb.org/t/p/",
      poster_sizes: ["w92", "w154", "w185", "w342", "w500", "w780"],
    },
  };
};

export const WithProvider = (props) => {
  const [state, dispatch] = useReducer(
    reducer,
    props.defaultValue || getTestState()
  );

  return (
    <AppContext.Provider
      value={{
        state,
        dispatch: props.mockDispatch || dispatch,
      }}
    >
      {props.children}
    </AppContext.Provider>
  );
};
