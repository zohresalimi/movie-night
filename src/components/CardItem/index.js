import React, { useState, useContext, useEffect } from "react";
import axiosInstance from "../../services";
import {
  Card,
  Image,
  Dimmer,
  Button,
  Message,
  Transition,
  Modal,
  Embed,
} from "semantic-ui-react";

import AppContext from "../../store/context";
import PlayButton from "../PlayButton";

import {
  SET_TO_FAVORITE_LIST,
  REMOVE_FROM_FAVORITE_LIST,
  SET_TO_WATCHLATER_LIST,
  REMOVE_FROM_WATCHLATER_LIST,
  SET_VIDEO_SOURCE,
} from "../../constants";

import "./style.css";

export default function CardItem({ item }) {
  const { state, dispatch } = useContext(AppContext);
  const { favoriteList, watchLaterList } = state;
  const [playVideo, setPlayVideo] = useState(false);
  const [isFavorite, setIsFavorite] = useState(() => !!favoriteList[item.id]);
  const [isSaved, setIsSaved] = useState(() => !!watchLaterList[item.id]);
  const [messageVisible, setMessageVisible] = useState(false);
  const [message, setMesage] = useState("");
  const [showDetail, setShowDetail] = useState(true);

  const [active, setactive] = useState();
  const { apiConfig } = state;

  const movieTrailerUrl = process.env.REACT_APP_GET_MOVIE_TRAILER_URL;

  const ToggleFavoriteList = () => {
    if (!favoriteList[item.id]) {
      dispatch({ type: SET_TO_FAVORITE_LIST, data: item });
      showMessage("Movie Added to the Favorit List");
    } else {
      dispatch({ type: REMOVE_FROM_FAVORITE_LIST, data: item });
      showMessage("Movie removed from the Favorit List");
    }
    setIsFavorite(!isFavorite);
  };

  const ToggleWathList = () => {
    if (!watchLaterList[item.id]) {
      dispatch({ type: SET_TO_WATCHLATER_LIST, data: item });
      showMessage("Movie Added to the Watch Later List");
    } else {
      dispatch({ type: REMOVE_FROM_WATCHLATER_LIST, data: item });
      showMessage("Movie removed from the Watch Later");
    }
    setIsSaved(!isSaved);
  };

  const handleDismiss = () => {
    setTimeout(() => {
      setMessageVisible(false);
    }, 2000);
  };

  const showMessage = (text) => {
    setMesage(text);
    setMessageVisible(true);
    handleDismiss();
  };

  useEffect(() => {
    const fetchMovieTrailer = async () => {
      debugger;
      try {
        const result = await axiosInstance.get(
          `${movieTrailerUrl}/${item.id}/videos`
        );
        console.log(result);
        if (result && result.data.results.length) {
          dispatch({
            type: SET_VIDEO_SOURCE,
            data: { result: result.data.results[0], selected: item },
          });
          setPlayVideo(true);
          setShowDetail(true);
        } else {
          setShowDetail(false);
        }
      } catch (err) {
        console.log(err);
      }
    };
    if (playVideo && !item.trailerKey) {
      fetchMovieTrailer();
    }
  }, [dispatch, item, movieTrailerUrl, playVideo]);

  const content = (
    <div>
      <PlayButton setPlayVideo={setPlayVideo} />
      <div className="tools-wrapper">
        <Button.Group vertical basic size="small">
          <Button
            icon="star"
            color={isFavorite ? "yellow" : ""}
            basic
            onClick={ToggleFavoriteList}
          />
          <Button
            icon="clock"
            color={isSaved ? "yellow" : ""}
            basic
            onClick={ToggleWathList}
          />
        </Button.Group>
      </div>
    </div>
  );
  return (
    <div className="card-wrapper">
      <Card>
        <Dimmer.Dimmable
          as={Image}
          dimmed={active}
          dimmer={{ active, content }}
          onMouseEnter={() => setactive(true)}
          onMouseLeave={() => setactive(false)}
          ui={false}
          wrapped
          size="small"
          src={`${apiConfig.secure_base_url}${apiConfig.poster_sizes[4]}/${item.poster_path}`}
        />
        <Card.Content>
          <Card.Header>{item.title}</Card.Header>
        </Card.Content>
      </Card>
      {messageVisible && (
        <Transition visible={messageVisible} animation="fade" duration={500}>
          <Message color="blue" size="huge" content={message} />
        </Transition>
      )}

      <Modal
        basic
        onClose={() => setPlayVideo(false)}
        open={playVideo}
        size="small"
      >
        <Modal.Content>
          {showDetail ? (
            <Embed
              id={item.trailerKey}
              autoplay
              placeholder={`${apiConfig.secure_base_url}${apiConfig.poster_sizes[4]}/${item.backdrop_path}`}
              source={item.trailerSite}
            />
          ) : (
            <Message
              negative
              icon="search"
              header="No Reasults"
              content="There is no Trailer for this Movie."
            />
          )}
        </Modal.Content>
      </Modal>
    </div>
  );
}
