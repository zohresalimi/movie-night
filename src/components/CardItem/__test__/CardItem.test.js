import React from "react";
import { render, act } from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect";

import { WithProvider } from "../../../mockTestData/data";
import CardItem from "../";

async function renderWrapper(props) {
  let component;

  act(() => {
    component = render(
      <WithProvider {...props}>
        <CardItem item={{ id: "" }} />
      </WithProvider>
    );
  });

  return component;
}

describe("CardItem Component Testing", () => {
  test("take snapshot", async () => {
    const { container } = await renderWrapper();
    expect(container).toMatchSnapshot();
  });
});
