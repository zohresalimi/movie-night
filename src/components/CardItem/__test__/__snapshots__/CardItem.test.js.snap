// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CardItem Component Testing take snapshot 1`] = `
<div>
  <div
    class="card-wrapper"
  >
    <div
      class="ui card"
    >
      <div
        class="small image dimmable"
      >
        <div
          class="ui dimmer"
        >
          <div
            class="content"
          >
            <div>
              <button
                data-testid="play-btn"
              >
                <a
                  class="sc-bdfBwQ gaGgLj"
                >
                  <i
                    aria-hidden="true"
                    class="play mini icon"
                  />
                </a>
              </button>
              <div
                class="tools-wrapper"
              >
                <div
                  class="ui small basic vertical buttons"
                >
                  <button
                    class="ui grey basic icon button"
                  >
                    <i
                      aria-hidden="true"
                      class="star icon"
                    />
                  </button>
                  <button
                    class="ui grey basic icon button"
                  >
                    <i
                      aria-hidden="true"
                      class="clock icon"
                    />
                  </button>
                </div>
              </div>
            </div>
          </div>
        </div>
        <img
          src="default-image.png"
        />
      </div>
      <div
        class="content"
      >
        <div
          class="header"
        />
      </div>
    </div>
  </div>
</div>
`;
